hail/python/hailtop/batch/backend.py

```diff
@@ -452,6 +459,26 @@ def __init__(self,
         gcs_kwargs = {'project': google_project}
         self.__fs: RouterAsyncFS = RouterAsyncFS(default_scheme='file', gcs_kwargs=gcs_kwargs)
 
+        self.regions: Optional[List[str]]
```

Comment:
None is no preference. [] means no regions are selected. We should not allow the empty list at all. I'll make the change across the board as soon as you give me the go-ahead to keep None as I had it.

Reply:
I decided to keep regions `Optional` here to signify that None means all available regions. However, in the driver code, the None from the JSON query that aggregates the region selections is converted to the list of all possible regions before being used in choosing a location.

Reply:
I commented about this above. I love using `Optional[X]` to ratchet on a unique new value to `X`, in this case to add a "the set of all regions" value. However, I also can see how this is confusing for users because "None" sounds like "No regions".

What do you think of this in hailtop.batch?

```
from typing import Union, List
from enum import Enum

class _ANY_REGION(Enum):
    ANY_REGION = 1


ANY_REGION = _ANY_REGION.ANY_REGION

REGION_SPECIFICATION = Union[_ANY_REGION, List[str]]

```

Reply:
Enum is just a way to get a singleton value with a type for mypy.

Reply:
We can add this if it makes sense to you. It is an extra concept and will be harder to document. Personally, I think just explaining what None means in the documentation is sufficient, but happy to go either way on this.